packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,252 @@
+                                      _dest_counts, _dest_offsets );
+
+        {
+            int const reorder = 0; // Ranking may *not* be reordered
+            int const sources[1] = {comm_rank};
+            int const degrees[1] = {static_cast<int>( destinations.size() )};
+
+            MPI_Dist_graph_create(
+                _comm, 1, sources, degrees, notNullPtr( destinations.data() ),
+                MPI_UNWEIGHTED, MPI_INFO_NULL, reorder, &_comm_dist_graph );
+        }
+
+        int indegrees;
+        int outdegrees;
+        int weighted;
+        MPI_Dist_graph_neighbors_count( _comm_dist_graph, &indegrees,
+                                        &outdegrees, &weighted );
+        DTK_ENSURE( weighted == 0 );
+        DTK_ENSURE( outdegrees == static_cast<int>( destinations.size() ) );
+
+        std::vector<int> sources( indegrees );
+        MPI_Dist_graph_neighbors( _comm_dist_graph, indegrees,
+                                  notNullPtr( sources.data() ), MPI_UNWEIGHTED,
+                                  outdegrees, notNullPtr( destinations.data() ),
+                                  MPI_UNWEIGHTED );
+
+        _src_counts.resize( indegrees );
+        MPI_Neighbor_alltoall( _dest_counts.data(), 1, MPI_INT,
+                               _src_counts.data(), 1, MPI_INT,
+                               _comm_dist_graph );
+
+        _src_offsets.resize( indegrees + 1 );
+        // exclusive scan
+        _src_offsets[0] = 0;
+        for ( int i = 0; i < indegrees; ++i )
+            _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+
+#if defined( REORDER_RECV )
+        int comm_size = -1;
+        MPI_Comm_size( _comm, &comm_size );
+        _permute_recv.resize( _src_offsets.back() );
+        int offset = 0;
+        for ( int i = 0; i < comm_size; ++i )
+        {
+            auto const it = std::find( sources.begin(), sources.end(), i );
+            if ( it != sources.end() )
+            {
+                int const j = std::distance( sources.begin(), it );
+                std::iota( &_permute_recv[offset],
+                           &_permute_recv[offset] + _src_counts[j],
+                           _src_offsets[j] );
+                offset += _src_counts[j];
+            }
+        }
+        DTK_ENSURE( offset == static_cast<int>( _permute_recv.size() ) );
+#endif
+
+        return _src_offsets.back();
+    }
+    template <typename View>
+    void doPostsAndWaits( typename View::const_type const &exports,
+                          size_t num_packets, View const &imports ) const
+    {
+        DTK_REQUIRE( num_packets * _src_offsets.back() == imports.size() );
+        DTK_REQUIRE( num_packets * _dest_offsets.back() == exports.size() );
+
+        using ValueType = typename View::value_type;
+        static_assert( View::rank == 1, "" );
+        static_assert(
+            Kokkos::Impl::MemorySpaceAccess<typename View::memory_space,
```

Comment:
So does this assume that when this is used elsewhere in the code that we will always have moved data to the host before this class is called?

Reply:
It does not assume anything.  It will not compile if the data is not accessible from the host.  I do not intend anyone to call `Details::Distributor::doPostsAndWaits()` directly, instead one should be using `sendAcrossNetwork(distributor, export, import)`.  I intend to change the interface of the `Distributor` and "merge" these two functions into a new member function.  It is beside the scope of this PR though.